### ISSUE TITLE
Support for torch.func

### ISSUE THREAD

**Issue by nicolas-dufour** [Lightning-AI/lightning#17820]
### Description & Motivation

Torch.func offers lots of flexibility. It's however not supported by lightning.
A few bugs I've noticed so far when trying to use it:

- When we init our model and then use a trainer, if we have used model.to("meta") to only keep the function version, it then fails as the trainers try to send it to device
- Params are not sent to the device, which then create device issues

Minimal example:

```python
class Module(L.LightningModule):
    def __init__(self, cfg):
        super().__init__()
        models = [torch.nn.Linear(in_features, out_features) for i in range(num_models)]
        self.fmodel = copy.deepcopy(models[0])
        self.model_params, self.model_buffers = torch.func.stack_module_state(models)

   def training_step(self, batch, batch_idx):
       # It is possible to vmap directly over torch.func.functional_call,
      # but wrapping it in a function makes it clearer what is going on.
      def call_single_model(params, buffers, data):
          return torch.func.functional_call(base_model, (params, buffers), (data,))
      output = torch.vmap(call_single_model, (0, 0, None))(self.model_params,  self.model_buffers, batch)
      return output.mean()
```

**Comment by ManoBharathi93** [comment]
@nicolas-dufour  As you noted, torch.func is a flexible tool for defining neural network models. However, it is not officially supported by PyTorch Lightning, and can lead to issues when trying to use it in conjunction with Lightning's device management or other features.

Approach:
One alternative that may be worth exploring is to use native PyTorch functions and modules instead of torch.func. In the specific example you provided, this would involve replacing the call to torch.func.functional_call with a standard forward pass through a linear layer using torch.functional.F.linear.

Example code:

```
def forward(self, x: torch.Tensor) -> torch.Tensor:
        def call_single_model(params, buffers, data):
            return torch.functional.F.linear(data, params, buffers)
```

**Comment by nicolas-dufour** [comment]
HI @ManoBharathi93 ,
I gave this example as a minimal reproduction example, but torch.func is very useful for a lot of applications, such as ensembles, complicated gradients computations among others. It's also core to some libraries such as torchrl.
Supporting it would help Lightning be more flexible and support more applications

**Comment by ManoBharathi93** [comment]
@nicolas-dufour  I completely understand your point! Torch.func is indeed a valuable feature for many applications. This would ultimately benefit users who rely on such functionality and enable them to take advantage of the full capabilities of both Lightning and PyTorch. If you haven't already, I would recommend discussing this further with the Lightning community to get their perspective and insights on how best to move forward with adding support for torch.func.
